Refactor server around configuration Observable

gabro opened the pull request:
This turned out to be a bit larger than I imagined, so I'm opening the PR although it's not completely done.

Some highlights:
- we now produce an `Observable[Configuration]` that gets passed around to providers
- as a result, providers are now in charge of instantiating whatever resource they need, using the configuration (e.g.  `DocumentFormattingProvider` creates a `Formatter`, instead of receiving it as input). I think this makes more sense in term of ownership and the `ScalametaLanguageServer` is now only managing providers, instead of all their dependencies)
- since now most providers' actions require accessing the configuration via `lastL`, they now return a `Task` directly, which again I think it makes sense. _"Task all the things!"_ (cit)
- I've decided to use `lastL` and not `lastOptionL`, which is a bit inconvenient to work with. In order to guarantee that the configuration observable is never empty I've decided to feed it a default configuration upon creation.

Future work (for another PR):
- finish adding features to the configuration
- use the features to dynamically register/unregister server capabilities (instead of determining them statically at boot)

### TODO
- [x] investigate test failures
- [x] add new configuration options to vscode client

gabro commented on metaserver/src/main/scala/scala/meta/languageserver/Formatter.scala:
```diff
@@ -19,7 +19,7 @@ abstract class Formatter {
 
   /** Returns formatter that does nothing */
-  def noop: Formatter = new Formatter {
+  lazy val noop: Formatter = new Formatter {
```
this isn't strictly related to this PR, but I think it makes sense to save some allocations

gabro commented on metaserver/src/main/scala/scala/meta/languageserver/ScalametaLanguageServer.scala:
```diff
@@ -435,6 +414,18 @@ object ScalametaLanguageServer extends LazyLogging {
+      .map(json => (json \ "scalameta").as[Configuration])
+      .doOnNext(conf => logger.info(s"Configuration updated $conf"))
+      .doOnError(e => connection.showMessage(MessageType.Error, e.getMessage))
+    subscriber.onNext(Json.toJson(Configuration()))
```
this is where I bootstrap the configuration, to make sure the observable is never empty. Not super elegant, but it works.

gabro commented on metaserver/src/main/scala/scala/meta/languageserver/compiler/ScalacProvider.scala:
```diff
@@ -16,9 +15,7 @@ import org.langmeta.inputs.Input
 
 /** Responsible for keeping fresh scalac global instances. */
-class ScalacProvider(
-    serverConfig: ServerConfig
```
`serverConfig` was never used.

olafurpg replied:
Probably relic from the time we indexed the classpath in ScalacProvider

ShaneDelmore commented on metaserver/src/main/scala/scala/meta/languageserver/Main.scala:
```diff
@@ -13,10 +13,11 @@ import org.langmeta.internal.io.PathIO
+    val logFile = configDir.resolve("metaserver.log").toFile
+    Files.createDirectories(configDir)
+    val out = new PrintStream(new FileOutputStream(logFile))
+    val err = new PrintStream(new FileOutputStream(logFile))
```
Is this ok to have two streams to the same file?  I have never tried it,  but would have expected err and out to point to the same stream, instead of having two streams point to the same file.

gabro replied:
This was already here, but it's a good question. I guess it's ok since it never gave us problems.

ShaneDelmore commented on metaserver/src/main/scala/scala/meta/languageserver/ScalametaLanguageServer.scala:
```diff
@@ -103,19 +90,19 @@ class ScalametaLanguageServer(
       OverflowStrategy.DropOld(2)
     )
+  val (configurationSubscriber, configurationPublisher) =
+    ScalametaLanguageServer.configurationStream(connection)
```
Why do we have connection here as well as lspIn and lspOut passed in?  I thought connection was just a wrapper around lspIn and lspOut.

gabro replied:
connection has higher level methods such as `showMessage`.

`lspIn` and `lspOut` are just used to initialize it (via the `LanguageServer` constructor)

ShaneDelmore commented on metaserver/src/main/scala/scala/meta/languageserver/ScalametaLanguageServer.scala:
```diff
@@ -228,20 +213,14 @@ class ScalametaLanguageServer(
-    }
-  }
+  override def onChangeConfiguration(settings: JsValue): Unit =
+    configurationSubscriber.onNext(settings)
```
What does this actually do?  I poked around the code and couldn’t figure it out.

gabro replied:
`onChangeConfiguration` is invoked whenever the client updates the settings. This line publishes the new value onto the Observable.

ShaneDelmore replied:
Just tracked that down.  Thanks for the confirmation.

gabro replied:
Ideally we will change the implementation to directly use Observables for notifications.

I.e. instead of having callbacks, we would just subscribe to the appropriate observables.

ShaneDelmore replied:
👍

ShaneDelmore commented on metaserver/src/main/scala/scala/meta/languageserver/providers/DocumentFormattingProvider.scala:
```diff
@@ -2,25 +2,40 @@ package scala.meta.languageserver.providers
+    else Formatter.noop
+  }
+
+  private def scalafmtConfigFromConfiguration: Task[AbsolutePath] = configuration.lastL.map { conf =>
```
This is just fine, but for general code review feedback, I would prefer to keep concepts together than are not useful separately. It feels to me like the missing concept here is that we just want a stream of format functions so we can always use the most recent.  That might be more clear if we just supplied an Observable[Formatter] or even Observable[FormattingFunction]

```scala
class DocumentFormattingProvider(configuration: Observable[Configuration], cwd: AbsolutePath) extends LazyLogging {

  private lazy val scalafmt = Formatter.classloadScalafmt("1.3.0")
  
  type FormatFunction = (String, String) => String
  val noFormat: FormatFunction = (str, _) => str
  val configuredFormatters: Observable[FormatFunction] = configuration.map{ conf =>
    if (conf.scalafmt.enable) {
      val fmtConf = cwd.resolve(conf.scalafmt.confPath)
      scalafmt.format(_, _, fmtConf)
    }
    else noFormat
  }
  
  def format(
      input: Input.VirtualFile
  ): Task[DocumentFormattingResult] =
    configuredFormatters.lastL.map{ fmt =>
          val fullDocumentRange = Range(
            start = Position(0, 0),
            end = Position(Int.MaxValue, Int.MaxValue)
          )
          val edits = {
            val formattedContent = fmt(input.value, input.path)
            List(TextEdit(fullDocumentRange, formattedContent))
          }
          DocumentFormattingResult(edits)
    }
}
```

gabro replied:
ah, I see. That's a very good point, and I actually had some doubts while making this refactor.
It's admittedly biased by the previous implementation, so I opted for the "minimum distance" change.
I'll give your feedback more thoughts later today, as I'm currently heading out right now.

ShaneDelmore replied:
I like small step refactors as well and as long as it is an improvement (which I say this is) I like to pull it in, we can always do another round of refactorings later.

olafurpg replied:
@ShaneDelmore `Formatter` is your `FormatFunction` 😜 I personally prefer methods + named parameters over higher-order functions on tuples. The former are easier to read, debug, use in the IDE, ...

ShaneDelmore replied:
I tend to prefer streams with functions applied to them (especially over callbacks), but I will keep try to keep your preferences in mind.

ShaneDelmore replied:
@olafurpg FWIW the most important thing to me usually is dealing in whole concepts when possible, position vs (line, column) and here my preferences is not a formatter function in general, but to get an observable of formatter, vs an unconfigured formatter and a format conf that need to be assembled before use when all anyone really wants is the thing you call to format the text.

ShaneDelmore commented on metaserver/src/test/scala/tests/compiler/SquiggliesTest.scala:
```diff
@@ -25,7 +34,7 @@ object SquiggliesTest extends CompilerSuite {
       val doc = Semanticdbs.toSemanticdb(input, compiler)
       val PublishDiagnostics(_, diagnostics) :: Nil =
-        SquiggliesProvider.squigglies(doc, linter)
+        squiggliesProvider.squigglies(doc).runSyncMaybe.right.get
```
Test helpers would be nice here so we didn't have to encode patterns in our tests that we wouldn't use in our code like .right.get.

For example:
```scala
  def assertNoDiffRight(task: Task[String], expected: String) = {
    val taskResult = task.runSyncMaybe
    taskResult match {
      case Left(cause) => fail("task failed...wtf?")
      case Right(value) => assertNoDiff(value, expected)
    }
  }

  def check(name: String, original: String, expected: String): Unit = {
    test(name) {
      val input = Input.VirtualFile(name, original)
      val doc = Semanticdbs.toSemanticdb(input, compiler)
      val obtained = squiggliesProvider.squigglies(doc)
        .map{ pubDiag =>
          val diags = pubDiag.flatMap(_.diagnostics)
          diags.map { d: =>
            val pos = input.toPosition(d.range)
            pos.formatMessage(d.severity.getOrElse(???).toString, d.message)
          }
        }

      assertNoDiffRight(obtained.map(_.mkString("\n")), expected)
    }
  }
```

olafurpg replied:
For me it's most important that errors don't get lost, `Try[T].get` is better than `Either[Err, T].right.get` in this regard. For Either I typically do `val Right(foo) = blah` instead of `right.get` for this exact reason, at least you get the left value on match error

ShaneDelmore replied:
Agreed.  That’s why rightTest or something like it is nice, it lets you make sure you always fail a test when you don’t get the right you want.  An anti-pattern I have seen (but not in this codebase) is developers doing SomeMonad.map(aTest) and having it fail silently.

olafurpg commented on metaserver/src/main/scala/scala/meta/languageserver/compiler/ScalacProvider.scala:
```diff
@@ -16,9 +15,7 @@ import org.langmeta.inputs.Input
-class ScalacProvider(
-    serverConfig: ServerConfig
-) extends LazyLogging {
+object ScalacProvider extends LazyLogging {
```
I'm not a fan of keeping state in objects, I prefer to `new X` to handle the lifecycle instead of leaving it to the JVM.

gabro replied:
Sure, I can make this a class with no parameters. It's probably a good idea since I suspect we may need to configure this provider too (e.g. with the version of Scala)

ShaneDelmore replied:
Good catch @olafurpg I missed that but agree, especially from a testing standpoint.

gabro replied:
done!

olafurpg commented on metaserver/src/main/scala/scala/meta/languageserver/providers/DocumentFormattingProvider.scala:
```diff
@@ -2,25 +2,40 @@ package scala.meta.languageserver.providers
+      input: Input.VirtualFile
+  ): Task[DocumentFormattingResult] = for {
+    formatter <- formatterFromConfiguration
+    config <- scalafmtConfigFromConfiguration
```
`Task[A].zip(Task[B])`?

gabro replied:
I find forcomprehension more readable and scalable than tupling things up. Unless there's a strong penalty in flatMapping over tasks

olafurpg replied:
for comprehension is fine, I agree it's more readable, wish we had ApplicativeDo in Scala 😏 https://www.microsoft.com/en-us/research/publication/desugaring-haskells-do-notation-into-applicative-operations/

ShaneDelmore replied:
I agree it more easily conveys the fact that I want to call A, then once A has returned call B.  But given that’s not what we want here, I think @olafurpg  version does convey the fact that these two things have no dependency on each other well.  So....sort of agree 🤷‍♂️

gabro replied:
Done, I've used zip.

olafurpg commented on metaserver/src/test/scala/tests/compiler/SquiggliesTest.scala:
```diff
@@ -25,7 +34,7 @@ object SquiggliesTest extends CompilerSuite {
       val doc = Semanticdbs.toSemanticdb(input, compiler)
       val PublishDiagnostics(_, diagnostics) :: Nil =
-        SquiggliesProvider.squigglies(doc, linter)
+        squiggliesProvider.squigglies(doc).runSyncMaybe.right.get
```
BTW there is a TestScheduler() which can run in sync mode with `tickOne()`, see how it's used in SymbolIndexTest

gabro replied:
I remembered you mentioning something like this a few days back, but I got lazy :D
Thanks for the pointer, I'll refactor this into something more appropriate

gabro replied:
I've kept `runSyncMaybe` but matched over `Right` on the lhs. I'm not sure I understand how the `TestScheduler` would help in this case, tbh.

olafurpg commented on metaserver/src/main/scala/scala/meta/languageserver/providers/DocumentFormattingProvider.scala:
```diff
@@ -2,25 +2,40 @@ package scala.meta.languageserver.providers
+  private lazy val scalafmt = Formatter.classloadScalafmt("1.3.0")
+
+  private def formatterFromConfiguration: Task[Formatter] = configuration.lastL.map { conf =>
+    if (conf.scalafmt.enable) scalafmt
```
I opened https://github.com/scalameta/language-server/issues/159 to track making the scalafmt version configurable. With `configuration.distinctByKey(_.scalafmt)` we can avoid re-classloading on every config change.

olafurpg commented on metaserver/src/main/scala/scala/meta/languageserver/Configuration.scala:
```diff
@@ -1,12 +1,73 @@
-    enableCompletions: Boolean,
+  scalafmt: Scalafmt = Scalafmt(),
+  scalafix: Scalafix = Scalafix(),
+  indexing: Indexing = Indexing(),
```
`s/indexing/search/`? With fields `indexJDK/indexClasspath`. Other fields that come to mind here are parameters for fuzzy search etc.

olafurpg commented on metaserver/src/main/scala/scala/meta/languageserver/Configuration.scala:
```diff
@@ -1,12 +1,73 @@
+
+  case class Experimental(
+      completions: Completions = Completions(),
+      hover: Hover = Hover(),
```
I think we can move hover and rename out of Experimental since both are built on semanticdb, esp. with edit distance we have quite a good story to reduce bogus results. The only part I'm not convinced about enabling by default is building semanticbs on the fly with the presentation compiler and completions.

olafurpg commented on metaserver/src/main/scala/scala/meta/languageserver/Configuration.scala:
```diff
@@ -1,12 +1,73 @@
 object Configuration {
+
+  case class Experimental(
+      completions: Completions = Completions(),
```
`s/completions/scalac/`? Feels consistent with scalafmt/scalafix, and if we decide to add other fields like sbt/scalariform/scalastyle

laughedelic commented on metaserver/src/main/scala/scala/meta/languageserver/Configuration.scala:
```diff
@@ -1,12 +1,73 @@
+      hover: Hover = Hover(),
+      rename: Rename = Rename()
+  )
+  case class Completions(enable: Boolean = false)
```
`config.completions.enable` reads like an action, would it be more appropriate to call it `enabled`?

gabro replied:
makes sense, done!

gabro commented:
Thank you all for the review! I've updated the current PR and also rebased on the latest master.
If we all agree on the current configuration, I'll add it to the vscode extension client, then we can merge this.

olafurpg commented on metaserver/src/main/scala/scala/meta/languageserver/ScalametaLanguageServer.scala:
```diff
@@ -449,6 +442,18 @@ object ScalametaLanguageServer extends LazyLogging {
+    val configurationPublisher = publisher
+      .map(json => (json \ "scalameta").as[Configuration])
+      .doOnNext(conf => logger.info(s"Configuration updated $conf"))
+      .doOnError(e => connection.showMessage(MessageType.Error, e.getMessage))
```
I think Observable stops on the first error so this is not enough to continue streaming configuration. I tried
```scala
    import monix.execution.Scheduler.Implicits.global
    val o = Observable
      .fromIterable(1.to(10))
      .map {
        case 5 => sys.error("5")
        case n => n + 1
      }
      .doOnError(e => println(s"EREEEEEEEEEE $e"))
    o.foreach(println)
```
and it prints out

```
2
3
4
5
EREEEEEEEEEE java.lang.RuntimeException: 5
```

I think it's best to avoid exceptions and do something like this instead
```scala
    import monix.execution.Scheduler.Implicits.global
    val o = Observable
      .fromIterable(1.to(10))
      .map {
        case 5 => Left(5)
        case n => Right(n)
      }
      .doOnNext {
        case Left(err) => println(s"FAIL $err")
        case _ => ()
      }
      .collect {
        case Right(x) => x
      }
    o.foreach(println)
```

which prints
```
1
2
3
4
FAIL 5
6
7
8
9
10
```

Alternatively, 

```scala
    val o = Observable
      .fromIterable(1.to(10))
      .liftByOperator[Int] { out =>
        new Subscriber[Int] {
          override implicit def scheduler: Scheduler = out.scheduler
          override def onError(ex: Throwable): Unit = out.onError(ex)
          override def onComplete(): Unit = out.onComplete()
          override def onNext(elem: Int): Future[Ack] = elem match {
            case 5 =>
              println("FAIL 5")
              Ack.Continue
            case n =>
              out.onNext(n)
          }
        }
      }
    o.foreach(println)
```
which prints the same.

gabro replied:
very nice catch, thank you!

olafurpg commented:
I've btw fixed the compile errors in https://github.com/olafurpg/language-server/compare/use-configuration?expand=1 if you want to use that commit, wanted to play with this locally

gabro commented on metaserver/src/main/scala/scala/meta/languageserver/ScalametaLanguageServer.scala:
```diff
@@ -293,20 +274,32 @@ class ScalametaLanguageServer(
 
   override def hover(
       request: TextDocumentHoverRequest
+<<<<<<< HEAD
```
wooops, that's what happens when I work in a rush...

gabro commented:
Thanks again for the review @olafurpg, ready for take 3! I've also committed the vscode-configuration, so now it works like

![](http://g.recordit.co/ekO53rrpKC.gif)

gabro commented:
The only bit I'm not sure of is why the success message is logged twice 🤔 Not a big deal, but there could be something wrong in the bindings

olafurpg commented on metaserver/src/main/scala/scala/meta/languageserver/providers/DocumentFormattingProvider.scala:
```diff
@@ -2,31 +2,52 @@ package scala.meta.languageserver.providers
+  private lazy val scalafmt = Formatter.classloadScalafmt("1.3.0")
+
+  private def formatterFromConfiguration: Task[Formatter] =
+    configuration.lastL.map { conf =>
```
I don't think `lastL` works as we expect it to 🤔 It seems to wait until the Observable completes before completing the Task.

```scala
    import monix.execution.Scheduler.Implicits.global
    import scala.concurrent.duration._
    val o = Observable
      .fromIterable(1.to(10))
      .delayOnNext(500.milliseconds)
    println("STARTING")
    val c = global.scheduleAtFixedRate(Duration.Zero, 1.second) {
      println("Tick")
      o.lastF.foreach(i => println(s"LAST: $i"))
    }
    o.dump("").subscribe()
    Thread.sleep(1000 * 10)
    c.cancel()
```

prints out
```
STARTING
Tick
0: -->1
Tick
1: -->2
2: -->3
Tick
3: -->4
4: -->5
Tick
5: -->6
6: -->7
Tick
7: -->8
8: -->9
Tick
9: -->10
10:  completed
LAST: 10
Tick
LAST: 10
Tick
LAST: 10
Tick
LAST: 10
Tick
LAST: 10
LAST: 10
```

I digged a bit into the monix sources and it seems we need to use MulticastStrategy.Behavior and use `.take(1).lastL`

```scala
    import monix.execution.Scheduler.Implicits.global
    import scala.concurrent.duration._
    val (sub, pub2) = Observable.multicast[Int](MulticastStrategy.behavior(1))
    val pub = pub2.delayOnNext(500.milliseconds)

    var i = 0
    global.scheduleAtFixedRate(Duration.Zero, 500.milliseconds) {
      sub.onNext(i)
      i += 1
    }
    val c = global.scheduleAtFixedRate(Duration.Zero, 1.second) {
      println("Tick")
      pub.take(1).lastL.foreach(i => println(s"LAST: $i"))
    }
    pub.dump("").subscribe()
    Thread.sleep(1000 * 10)
    c.cancel()
```

Produces
```
Tick
0: -->0
LAST: 0
Tick
1: -->1
LAST: 1
2: -->2
Tick
3: -->3
LAST: 3
4: -->4
Tick
5: -->5
LAST: 5
6: -->6
Tick
7: -->7
LAST: 7
8: -->8
Tick
9: -->9
LAST: 9
10: -->10
Tick
11: -->11
LAST: 11
12: -->12
Tick
13: -->13
LAST: 13
14: -->14
Tick
15: -->15
LAST: 15
16: -->16
Tick
17: -->17
LAST: 17
18: -->18
19: -->19
```

Behavior takes an initial value which makes the initial `onNext` unnecessary. This reproduces "reactive variables" according to https://monix.io/docs/2x/reactive/observable-comparisons.html

ShaneDelmore replied:
Monix has built-ins for watching files in their monix-nio package.  https://github.com/monix/monix-nio/tree/master/src/main/scala/monix/nio/file

gabro replied:
@olafurpg thanks for digging into it. I've refactored it to use a `Behavior` strategy. I definitely need to study Monix a bit more to fully grasp these variants.

gabro commented on metaserver/src/main/scala/scala/meta/languageserver/ScalametaLanguageServer.scala:
```diff
@@ -434,9 +436,17 @@ object ScalametaLanguageServer extends LazyLogging {
-      .doOnNext(conf => logger.info(s"Configuration updated $conf"))
-      .doOnError(e => connection.showMessage(MessageType.Error, e.getMessage))
+      .map(json => (json \ "scalameta").validate[Configuration])
+      .doOnNext {
```
yeah, I felt the same. I've tried playing around with `onErrorHandle` and similar, but I couldn't find a combinator that suits it.

gabro commented on metaserver/src/main/scala/scala/meta/languageserver/providers/DocumentFormattingProvider.scala:
```diff
@@ -2,31 +2,52 @@ package scala.meta.languageserver.providers
+  private lazy val scalafmt = Formatter.classloadScalafmt("1.3.0")
+
+  private def formatterFromConfiguration: Task[Formatter] =
+    configuration.take(1).lastL.map { conf =>
```
this seems to work, but forgetting `take(1)` it's a very easy mistake to make. I wonder whether we're missing something more obvious.

gabro commented on metaserver/src/main/scala/scala/meta/languageserver/Configuration.scala:
```diff
@@ -1,12 +1,85 @@
+    scalafix: Scalafix = Scalafix(),
+    search: Search = Search(),
+    experimental: Experimental = Experimental(),
+    hover: Hover = Hover(),
```
#161

olafurpg commented:
I pushed a few minor fixes, but
```
00:11:24.640 INFO  s.m.l.ScalametaLanguageServer$ - Configuration updated Configuration(Scalafmt(true,.scalafmt.conf),Scalafix(true,.scalafix.conf),Search(false,true),Experimental(Scalac(false)),Hover(true),Rename(true))
```
still runs on every format, which is not ideal.

gabro commented:
Thanks for the fixes, thanks! I've been working on this on 30m increments over the last few days, and I've been scattering minor errors all over the place...

As per the open points:



> I'm also seeing this cryptic error appear while tweaking the settings
> 
> ```
> 23:27:39.493 ERROR l.c.LanguageServer$$anon$1 - Invalid Notification: List((,List(JsonValidationError(List(unknown method $/setTraceNotification),WrappedArray())))) - Message: JsonRpcNotificationMessage($/setTraceNotification,ObjectParams({"value":"off"}))
> ```

the requests starting with `$` can be safely ignored by the server, according to the spec. That specific message is specific of vscode-languageclient and we should disregard it. See https://github.com/Microsoft/vscode-languageserver-node/issues/170

The json-rpc implementation we're using is picky about unknown messages and it logs the error.
I think we can safely ignore it for now.

> It might be worth opening an issue to automatically generate the json commands from the defaults.

Absolutely. For vscode it should be easy: read the configuration either with Scalameta or directly from Scala and output a json document. I don't know how they are configured in other clients.

> Running "Format document" seems to re-log
> ```
> 23:19:51.634 INFO  s.m.l.ScalametaLanguageServer$ - Configuration updated Configuration(Scalafmt(false,.scalafmt.conf),Scalafix(true,.scalafix.conf),Search(false,true),Experimental(Scalac(false)),Hover(true),Rename(true))
> ```

That's surprising. Do you have any idea of why it happens? Could it be that `take(1)` triggers `onNext`?

gabro commented on metaserver/src/main/scala/scala/meta/languageserver/providers/DocumentFormattingProvider.scala:
```diff
@@ -1,52 +1,91 @@
+      .map[Either[String, Option[AbsolutePath]]] {
+        case None =>
+          val default = cwd.resolve(Scalafmt.DefaultConf)
+          if (Files.isRegularFile(default.toNIO)) Right(Some(default))
```
This can probably be a shared utility function:

```scala
def pathIfExists(path: AbsolutePath): Option[AbsolutePath] =
  Option(path).filter(p => Files.isRegularFile(p.toNIO))
```

gabro commented on metaserver/src/main/scala/scala/meta/languageserver/providers/SquiggliesProvider.scala:
```diff
@@ -6,31 +6,35 @@ import com.typesafe.scalalogging.LazyLogging
-      config <- configuration
-    } yield {
+  def squigglies(db: m.Database): Task[Seq[PublishDiagnostics]] = Task.eval {
+    if (!isEnabled()) Nil
```
`isEnabled` is based on `scalafix.enabled`. This check also skips the compiler errors, which are not related.

olafurpg commented:
I can't respond to most your review @gabro so I'll answer here instead

> let's change this also in the vscode extension, by omitting the defaultValue.

The default value is still .scalafmt.conf, I used Option to help report errors when the configured (non-empty) setting points to a file that doesn't exist.

> DefaultConf => defaultConfPath?

Done.

> I would leave enabled in, even if unused, in preparation for #163

Done. We should implement all settings before releasing v0.1, I'm a bit concerned about having unused settings.

Finally. I took the liberty to remove the Experimental case class. I personally thought it was a good idea before but now I think we should rather strive to never break configuration files in upgrades. I added "EXPERIMENTAL" to the setting description for disabled-by-default features.

olafurpg commented:
> We can replace the nested match using a for-comprehension and the valueOr from cats.

Done! I didn't use valueOr. With the ongoing lsp4s refactoring the signature for handling requests will be
```scala
  def handle(request: A): Task[Either[Response.Error, B]]
```
where we could return something like `Response.invalidParams` instead of `showMessage()` + return dummy value

gabro commented:
78 messages and 17 commits later... 🎉

olafurpg commented:
🎉
